### PR TITLE
fix(optimizer): Fix optimizer for generate series

### PR DESCRIPTION
**Pull request by chrisqu777**
### Issue
As described in #6657, when a subquery inside GENERATE_SERIES is unnested, the optimizer currently moves it to the end of the query plan. This prevents the function from seeing the referenced column/table, causing resolution errors. However, while testing query: `SELECT t1.c1 FROM z RIGHT JOIN GENERATE_SERIES((SELECT MIN(x.a) FROM x), 10, 1) AS t1(c1) ON t1.c1 > z.b;`, we found unnesting it will requires different handling approach depending on whether table `z` is empty or not, which causes further issue.

### Fix
Instead of putting the subquery before or behind `GENERATE_SERIES`, it makes more sense to avoid unnesting this subquery to keep sematic, so that the result would always be correct no matter `GENERATE_SERIES` joins with empty/non-empty table in each side. 
The fix adds the `skip_expressions` variable to explicitly define the expression to be skipped as well as the reason why we skip it.

### Test
Tests are added based on the findings when researching this issue, this would help avoid unexpected changes around this function.

### CODE REVIEW

File: sqlglot/optimizer/unnest_subqueries.py [hunk 1]
```diff
@@ -83,6 +83,35 @@ def unnest(select, parent_select, next_alias_name):
...
+            # appear before that function. Otherwise, the subquery reference
+            # used in the function will be invalid.
+            parent_select_from = parent_select.args.get("from_")
+            parent_select_joins = parent_select.args.get("joins")
+
+            parent_select_from = parent_select_from.this
+            unnested_subquery = parent_select_joins.pop().this
+
+            # Cross join the unnested subquery before the first reference
+            if any(col.table == alias for col in parent_select_from.find_all(exp.Column)):
```
Comment (VaggelisD) [inline sqlglot/optimizer/unnest_subqueries.py]: In case of `FROM GENERATE_SERIES(...)` this branch is equivalent to the following afaict:

```suggestion
            if parent_select_from is predicate.parent:
```

File: sqlglot/optimizer/unnest_subqueries.py [hunk 2]
```diff
@@ -83,6 +83,35 @@ def unnest(select, parent_select, next_alias_name):
...
+                parent_select.args["from_"].replace(
+                    exp.From(this=unnested_subquery, alias=unnested_subquery.alias)
+                )
+                parent_select.args["joins"] = [
+                    exp.Join(this=parent_select_from, kind="CROSS"),
+                    *parent_select_joins,
+                ]
+            else:
+                for i, join in enumerate(parent_select_joins):
+                    if any(col.table == alias for col in join.find_all(exp.Column)):
```
Comment (VaggelisD) [inline sqlglot/optimizer/unnest_subqueries.py]: We can simply find `exp.GenerateSeries` in the join list and append before it, right? No need to also tree walk into each `exp.Column` reference, which is expensive

File: sqlglot/optimizer/unnest_subqueries.py [hunk 3]
```diff
@@ -83,6 +83,35 @@ def unnest(select, parent_select, next_alias_name):
...
+                        parent_select.args["joins"] = [
+                            *parent_select_joins[:i],
+                            exp.Join(this=unnested_subquery, kind="CROSS"),
+                            *parent_select_joins[i:],
+                        ]
```
Comment (VaggelisD) [inline sqlglot/optimizer/unnest_subqueries.py]: Should we `insert` at `i - 1`? Might be easier to read

File: sqlglot/optimizer/unnest_subqueries.py [hunk 4]
```diff
@@ -83,6 +83,35 @@ def unnest(select, parent_select, next_alias_name):
...
+            parent_select_from = parent_select.args.get("from_")
+            parent_select_joins = parent_select.args.get("joins")
+
+            parent_select_from = parent_select_from.this
+            unnested_subquery = parent_select_joins.pop().this
+
+            # Cross join the unnested subquery before the first reference
+            if any(col.table == alias for col in parent_select_from.find_all(exp.Column)):
+                parent_select.args["from_"].replace(
+                    exp.From(this=unnested_subquery, alias=unnested_subquery.alias)
```
Comment (VaggelisD) [inline sqlglot/optimizer/unnest_subqueries.py]: `exp.From` only has a `this` arg, the alias is already embedded into the subquery

```suggestion
                    exp.From(this=unnested_subquery)
```

**Review by VaggelisD — COMMENTED** [review]
Thank you for tackling this @chrisqu777, nice work! Leaving a few comments:

File: sqlglot/optimizer/unnest_subqueries.py (hunk 3 above)
Comment (chrisqu777) [reply sqlglot/optimizer/unnest_subqueries.py]: Yes, we can use insert here.

File: sqlglot/optimizer/unnest_subqueries.py (hunk 2 above)
Comment (chrisqu777) [reply sqlglot/optimizer/unnest_subqueries.py]: @VaggelisD For this section, what do you think about replacing the expression in FROM with the subquery? 

Initially, I avoided placing the subquery at the beginning because I was concerned about potential side effects. However, since the GENERATE_SERIES subquery emits only a single value, testing several edge cases shows that putting it directly in the FROM clause produces the same result. In addition, doing so simplifies the code and makes it more readable.

Comment (chrisqu777) [reply sqlglot/optimizer/unnest_subqueries.py]: I've updated the code to show this. Let me know what you think.

Comment (VaggelisD) [reply sqlglot/optimizer/unnest_subqueries.py]: Hey @chrisqu777, I also had the same thought regarding side effects, the previous version was easier to reason about given that the join order was preserved.

Could you add the edge cases you tested as extra cases?

File: tests/fixtures/optimizer/optimizer.sql [hunk 5]
```diff
@@ -1536,3 +1536,16 @@ FROM "t1" AS "t1"
 RIGHT JOIN "t2" AS "t2"
   ON "t1"."id1" = "t2"."id2"
 CROSS JOIN "t3" AS "t3";
+
+# title: unnested subquery show before the reference
+# execute: false
```
Comment (VaggelisD) [inline tests/fixtures/optimizer/optimizer.sql:1541]: We had 2 tests before whereas now we have 1, can we add the previous test plus any case useful re: https://github.com/tobymao/sqlglot/pull/6679#discussion_r2682130901 ?

File: sqlglot/optimizer/unnest_subqueries.py [hunk 6]
```diff
@@ -83,6 +83,22 @@ def unnest(select, parent_select, next_alias_name):
...
+            parent_select_from = parent_select.args.get("from_")
+            parent_select_joins = parent_select.args.get("joins")
+
+            unnested_subquery = parent_select_joins.pop()
+
+            # The unnested subquery only emits constant value, so declare it first
+            parent_select.args["from_"].replace(exp.From(this=unnested_subquery.this))
+            parent_select.args["joins"].insert(
+                0, exp.Join(this=parent_select_from.this, kind="CROSS")
+            )
```
Comment (VaggelisD) [inline sqlglot/optimizer/unnest_subqueries.py]: I guess if `S` was the unnested subquery, then the fact that it produces only 1 value means that it should be fine to put it first in the join order since its `CROSS JOIN` simply "appends" a new column to the rest of the join result, right?

File: sqlglot/optimizer/unnest_subqueries.py [hunk 7]
```diff
@@ -83,6 +83,22 @@ def unnest(select, parent_select, next_alias_name):
 
         _replace(select.parent, column)
         parent_select.join(select, on=on_clause, join_type=join_type, join_alias=alias, copy=False)
+
+        if isinstance(predicate, exp.GenerateSeries):
```
Comment (georgesittas) [inline sqlglot/optimizer/unnest_subqueries.py]: Is this actually specific to `GenerateSeries`, though? Should this be something more generic, like "is this a func that appears in the `FROM` clause or a `JOIN`"?

File: sqlglot/optimizer/unnest_subqueries.py (hunk 6 above)
Comment (georgesittas) [reply sqlglot/optimizer/unnest_subqueries.py]: Does the `parent_select` always have a `joins` arg set? Could this fail abruptly?

Comment (VaggelisD) [reply sqlglot/optimizer/unnest_subqueries.py]: I think so, because a few lines above we've added the join only to `pop()` it later:

```Python3
parent_select.join(select, on=on_clause, join_type=join_type, join_alias=alias, copy=False)

if isinstance(predicate, exp.GenerateSeries):
  ...
  unnested_subquery = parent_select_joins.pop()
```

File: sqlglot/optimizer/unnest_subqueries.py (hunk 7 above)
Comment (chrisqu777) [reply sqlglot/optimizer/unnest_subqueries.py]: Currently, this issue occurs when a function contains a subquery that returns a single value. The subquery is extracted and appended as a CROSS JOIN at the end of the join list, which leads to the incorrect behavior.

By placing this logic under exp.GenerateSeries, the change is safer and more localized, and it also makes it easier to extend support for other expression types in the future if needed.

File: sqlglot/optimizer/unnest_subqueries.py (hunk 6 above)
Comment (chrisqu777) [reply sqlglot/optimizer/unnest_subqueries.py]: @georgesittas Yes, as this section happens after line 85: `parent_select.join(select, on=on_clause, join_type=join_type, join_alias=alias, copy=False)`, which starts a join.

File: tests/fixtures/optimizer/optimizer.sql (hunk 5 above)
Comment (chrisqu777) [reply tests/fixtures/optimizer/optimizer.sql:1541]: I’ll add tests for the edge cases. The previous test was intended to ensure that the subquery could CROSS JOIN with generate_series immediately after it. Since the subquery is now placed at the top, this is no longer necessary.

File: sqlglot/optimizer/unnest_subqueries.py (hunk 6 above)
Comment (georgesittas) [reply sqlglot/optimizer/unnest_subqueries.py]: Alright, let's change L91-L92 to also use the `[...]` notation instead of `get(...)`, then.

File: sqlglot/optimizer/unnest_subqueries.py (hunk 7 above)
Comment (georgesittas) [reply sqlglot/optimizer/unnest_subqueries.py]: Right, but my point is whether we should generalize this beyond `GenerateSeries`. For example, does the same issue occur when using a subquery within a BigQuery `UNNEST` operation?

```
(sqlglot) ➜  sqlglot git:(main) bq query "select * from unnest((select ['a', 'b', 'c'])) c"
+---+
| c |
+---+
| a |
| b |
| c |
+---+
```

So.. if we handle `Func`, which is the broader superclass, and perhaps combine it with a check "is this under a `FROM` or` JOIN`", that may fully fix this error class for BigQuery / others too (assuming the same problem exists for my example above).

File: sqlglot/optimizer/unnest_subqueries.py [hunk 8]
```diff
@@ -40,11 +40,16 @@ def unnest(select, parent_select, next_alias_name):
...
+    # Expressions that cannot be safely converted by unnesting
+    # exp.GenerateSeries: skipped as unnesting will change semantics if join with other tables.
+    skip_expressions = (exp.GenerateSeries,)
+
     predicate = select.find_ancestor(exp.Condition)
     if (
         not predicate
         or parent_select is not predicate.parent_select
         or not parent_select.args.get("from_")
+        or isinstance(predicate, skip_expressions)
```
Comment (VaggelisD) [inline sqlglot/optimizer/unnest_subqueries.py:52]: [Nit] We can bring this up to L50 in order to short-circuit early and not have to check anything parent related, e.g:

```Python3
  if (
        not predicate
        or isinstance(predicate, skip_expressions)
        or parent_select is not predicate.parent_select
        or not parent_select.args.get("from_")
    ):
        return
```

File: sqlglot/optimizer/unnest_subqueries.py [hunk 9]
```diff
@@ -40,11 +40,16 @@ def unnest(select, parent_select, next_alias_name):
     if len(select.selects) > 1:
         return
 
+    # Expressions that cannot be safely converted by unnesting
+    # exp.GenerateSeries: skipped as unnesting will change semantics if join with other tables.
+    skip_expressions = (exp.GenerateSeries,)
+
```
Comment (VaggelisD) [inline sqlglot/optimizer/unnest_subqueries.py:46]: [Nit] We can move this outside of the function, may also help to rename it to `UNNESTABLE_EXPRESSIONS` or something equivalent

**Comment by VaggelisD** [comment]
Will merge these and get the rest of the nits to the finish line, thanks again @chrisqu777!
